### ISSUE TITLE
Support for ARM platforms with Mali GPU (OpenCL)

### ISSUE THREAD

**Issue by Butakus** [arrayfire/arrayfire#2087]
First of all, thank you for the amount of great work you put in the arrayfire library.

I am trying to build arrayfire on the Odroid XU4, which is an armv7l platform with a Mali GPU. This device is compatible only with OpenCL 1.1, but current arrayfire master needs at least OpenCL 1.2 version to build.

As I have seen [here](https://arrayfire.com/arrayfire-write-once-run-anywhere/), it is possible to run arrayfire on ARM platforms with Mali GPU using OpenCL 1.1, though that page seems a bit old (from 2014).

Is the 32 bit ARM platform with Mali GPU still supported by the library? Would it be possible to have access to the version of the library mentioned in the link above?

Thank you in advance

**Comment by 9prady9** [comment]
@Butakus As you can see below and in the blog post you shared earlier,
```
ArrayFire v2.1  (OpenCL, 32-bit Linux, build 1971d38)
```
The ARM platform support was available in version 2.* albiet on-demand. It was experimental and the library binaries for said platforms were only provided on request by our customers. When we open sourced our library and moved to v3.0, we only ported the support for 64 bit AMD64`x86_64` architecture. The major reason for doing so is the support for OpenCL on mobile platforms is not wide.  Having said that, we still support arrayfire on one ARM platform - NVIDIA Tegra boards, we test them regularly on our servers too.

Technically, you might be able to build current master for Mali GPU or Adreno GPUs but as far as I can remember they will need further tuning to get good performance out of them. I haven't tried arrayfire v3 on any mobile GPU yet.

The bottom line is that we won't be able to get to adding ARM support for Mali/Adreno GPUs any time. Right now, we only support NVIDIA Tegra platform.

**Comment by Butakus** [comment]
> Technically, you might be able to build OpenCL backend on current master for Mali GPU or Adreno GPUs

I guess this is only possible if OpenCL 1.2 is installed. Since I can only have OpenCL 1.1, it won't be an option.


It makes sense that you cannot support it, and I understand perfectly. I will consider moving to the Tegra boards (which have the plus of CUDA) and you can close this issue if there are no more comments.

Thank you!

**Comment by 9prady9** [comment]
Most of our kernels don't use any fancy stuff, so you might be able to get it running with few tweaks. Although we can't exactly tell you what changes needed to be done to get it going.

Thank you for your interest! Do try out ArrayFire on NVIDIA tegra and raise questions on our [google forums](https://groups.google.com/forum/#!forum/arrayfire-users) and issues if you face any on github.
